src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,835 @@
+        for msg in input {
+            let msg = msg?;
+            match msg {
+                Message::Barrier(barrier) => {
+                    if let Some(ref mutation) = barrier.mutation.as_deref() {
+                        match mutation {
+                            Mutation::Pause | Mutation::Resume => {
+                                // We don't need to do anything. Handled by upstream.
+                            }
+                            Mutation::SourceChangeSplit(actor_splits) => {
+                                tracing::info!(
+                                    actor_id = self.actor_ctx.id,
+                                    actor_splits = ?actor_splits,
+                                    "source change split received"
+                                );
+                                self.apply_split_change_forward_stage(
+                                    actor_splits,
+                                    &mut splits,
+                                    true,
+                                )
+                                .await?;
+                            }
+                            Mutation::Update(UpdateMutation { actor_splits, .. }) => {
+                                self.apply_split_change_forward_stage(
+                                    actor_splits,
+                                    &mut splits,
+                                    true,
+                                )
+                                .await?;
+                            }
+                            _ => {}
+                        }
+                        self.backfill_state_store
+                            .state_store
+                            .commit(barrier.epoch)
+                            .await?;
+                    }
+                    yield Message::Barrier(barrier);
+                }
+                Message::Chunk(chunk) => {
+                    yield Message::Chunk(chunk);
+                }
+                Message::Watermark(watermark) => {
+                    yield Message::Watermark(watermark);
+                }
+            }
+        }
+    }
+
+    /// All splits finished backfilling.
+    ///
+    /// We check all splits for the source, including other actors' splits here, before going to the forward stage.
+    /// Otherwise if we break early, but after rescheduling, an unfinished split is migrated to
+    /// this actor, we still need to backfill it.
+    async fn backfill_finished(&self, states: &BackfillStates) -> StreamExecutorResult<bool> {
+        Ok(states
+            .values()
+            .all(|state| matches!(state, BackfillState::Finished))
+            && self
+                .backfill_state_store
+                .scan()
+                .await?
+                .into_iter()
+                .all(|state| matches!(state, BackfillState::Finished)))
+    }
+
+    /// For newly added splits, we do not need to backfill and can directly forward from upstream.
```

Comment:
Is it possble that adding splits (in Kafka cluster) just happens before the new upstream `SourceExecutor` created? So a few events at the beginning could be missed.

On the other hand, how about running backfilling for any new splits? Would it introduce any additional cost?

Reply:
> how about running backfilling for any new splits? Would it introduce any additional cost?

The main reason is for code structure. I didn't come up with a way to backfill new added splits after we enter the "forwarding stage". One solution is that we just remain inside the `backfill_loop`, and do not go to "forwarding stage" at all. Maybe the overhead isn't too large. But since source is perf-critical part...

Reply:
> Is it possble that adding splits (in Kafka cluster) just happens before the new upstream SourceExecutor created? So a few events at the beginning could be missed.

I don't get it. In this case, for all events that go through Source, they should also go through SourceBackfill, so no data are lost. 

IIUC it works like:
1. Source and SourceBackfill all starts with old splits.
2. Meta source manager detects newly added split and sends a barrier command.
3. Source and SourceBackfill does split change simultaneously.

Reply:
Convincing